src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,451 @@
+	const int max_weight = 10;
+	seek.sample_weights.resize(
+	        seek.samples.size());
+	for (unsigned int i = 0; i < seek.samples.size(); ++i) {
+		seek.sample_weights[i] = static_cast<int>(
+		        max_weight - i);
+	}
+}
+
+int DiskNoiseDevice::ChooseWeightedSeekIndex() const
+{
+	int total_weight = 0;
+	for (unsigned int i = 0; i < seek.samples.size(); ++i) {
+		if (!seek.samples[i].empty()) {
+			total_weight += seek.sample_weights[i];
+		}
+	}
+
+	if (total_weight == 0) {
+		return 0;
+	}
+
+	const int r = static_cast<int>(rand()) % total_weight;
+	int sum     = 0;
+	for (unsigned int i = 0; i < seek.samples.size(); ++i) {
+		if (seek.samples[i].empty()) {
+			continue;
+		}
+
+		sum += seek.sample_weights[i];
+		if (r < sum) {
+			return i;
+		}
+	}
+	return 0;
+}
+DiskNoiseDevice::DiskNoiseDevice(const DiskType disk_type,
+                                 const bool enable_disk_noise,
+                                 const std::string& spin_up_sample_path,
+                                 const std::string& spin_sample_path,
+                                 const std::vector<std::string>& seek_sample_paths,
+                                 bool loop_spin_sample)
+        : enable_disk_noise(enable_disk_noise)
+{
+	if (!enable_disk_noise) {
+		LOG_INFO("DISKNOISE: Disk noise emulation disabled");
+		return;
+	}
+	spin.loop = loop_spin_sample;
+	LoadSample(spin_up_sample_path, spin.spin_up_sample);
+	LoadSample(spin_sample_path, spin.sample);
+	LoadSeekSamples(seek_sample_paths);
+
+	disk_noises->active_devices.push_back(this);
+	DOS_RegisterIoCallback([this]() {
+		// This callback is called from the DOS code
+		// to trigger the spin and seek sounds
+		ActivateSpin();
+		PlaySeek();
+	}, disk_type);
+}
+
+void DiskNoiseDevice::ActivateSpin()
+{
+	if (!enable_disk_noise) {
+		return;
+	}
+
+	// Floppy spin samples can be re-started at any time
+	if (!spin.loop) {
+		// Check if the sample is still playing and don't interrupt if
+		// it does
+		if (spin.sample.empty() || spin.pos + 1 < spin.sample.size()) {
+			return;
+		}
+		// Restart spin sample
+		spin.pos = 0;
+	}
+}
+
+void DiskNoiseDevice::PlaySeek()
+{
+	if (!enable_disk_noise) {
+		return;
+	}
+
+	const size_t index = ChooseWeightedSeekIndex();
+	if (index >= seek.samples.size() ||
+	    seek.samples[index].empty()) {
+		return;
+	}
+
+	// Check if the sample is still playing and don't interrupt if it does
+	if (!seek.current_sample.empty() &&
+	    seek.pos + 1 < seek.current_sample.size()) {
+		return;
+	}
+
+	seek.current_sample = seek.samples[index];
+	seek.pos            = 0;
+}
+
+void DiskNoiseDevice::Shutdown()
+{
+	std::lock_guard<std::mutex> lock(disk_noises->device_mutex);
+	disk_noises->active_devices.erase(
+	        std::remove(disk_noises->active_devices.begin(),
+	                    disk_noises->active_devices.end(),
+	                    this),
+	        disk_noises->active_devices.end());
```

Comment:
Do this in the destructor, please. No two-stage init/shutdown if we can avoid it, and here we easily can.